labgrid/util/helper.py

```diff
@@ -26,21 +27,34 @@ def get_user():
     import pwd
     return pwd.getpwuid(os.getuid())[0]
 
+def set_nonblocking(fd):
+    flags = fcntl.fcntl(fd, fcntl.F_GETFL)
+    fcntl.fcntl(fd, fcntl.F_SETFL, flags | os.O_NONBLOCK)
+
 @attr.s
 class ProcessWrapper:
     callbacks = attr.ib(default=attr.Factory(list))
     loglevel = logging.INFO
 
     @step(args=['command'], result=True, tag='process')
-    def check_output(self, command, *, print_on_silent_log=False):
+    def check_output(self, command, *, print_on_silent_log=False, input=None): # pylint: disable=redefined-builtin
         """Run a command and supply the output to callback functions"""
         logger = logging.getLogger("Process")
         res = []
         mfd, sfd = pty.openpty()
-        flags = fcntl.fcntl(mfd, fcntl.F_GETFL)
-        fcntl.fcntl(mfd, fcntl.F_SETFL, flags | os.O_NONBLOCK)
+        set_nonblocking(mfd)
+
+        kwargs = {}
+
+        stdin_r = None
+        stdin_w = None
+        if input is not None:
+            stdin_r, stdin_w = os.pipe()
+            kwargs['stdin'] = stdin_r
```

Comment:
I'm probably missing something... Why can't we use the pty for stdin as well?

Reply:
In order to inject a string as stdin (which is what `input` does), you need to be able to independently close stdin after the input is consumed; think of a program like `cat` that terminates on input EOF. If you tied it into the the pty, you would also close stdout when you did that.

Reply:
Or put another way, `input` should be more like doing `echo <INPUT> | PROGRAM` rather than interactive (where I think you would do it through the pty).